i18n-helpers/src/bin/mdbook-xgettext.rs

```diff
@@ -86,21 +103,161 @@ fn create_catalog(ctx: &RenderContext) -> anyhow::Result<Catalog> {
+                        template_destination.clone(),
+                        Catalog::new(generate_catalog_metadata(&ctx)),
+                    );
+                }
+                let catalog = catalogs.get_mut(&template_destination).unwrap();
+                let source = ctx.config.book.src.join(&source);
+
+                for (lineno, msgid) in extract_messages(&content) {
+                    let source_with_lineno = format!("{}:{}", source.display(), lineno);
+                    add_message(catalog, &msgid, &source_with_lineno);
+                }
             }
         }
     }
 
-    Ok(catalog)
+    Ok(catalogs)
+}
+
+/// A view into the relevant template information held by `mdbook::book::Chapter`
+/// and a location to store the exported polib messages.
+struct ChapterTemplateInfo {
+    /// The chapter's content.
+    content: String,
+    /// The file where the content is sourced.
+    source: PathBuf,
+    /// The output destination for the polib template.
+    template_destination: PathBuf,
+}
+
+// A recursive function to crawl a chapter's sub-items and get
+// the relevant info to produce a set of po template files.
+fn get_subcontent_for_chapter(
+    c: &Chapter,
+    provided_file_path: PathBuf,
+    provided_depth: i64,
+    depth: i64,
+) -> Vec<ChapterTemplateInfo> {
+    if c.sub_items.is_empty() {
+        return Vec::new();
+    };
+
+    // Iterate through sub-chapters and pull the chapter content, path,
+    // and destination to store the template.
+    c.sub_items
+        .iter()
+        .filter_map(|item| {
+            if let BookItem::Chapter(chapter) = item {
+                let (chapter_info, new_path) = match &chapter.path {
+                    Some(chapter_path) => {
+                        // Append the chapter's name to the template's destination
+                        // when the depth has not surpassed the provided value.
+                        let destination = if depth < provided_depth {
+                            provided_file_path.join(slug(&chapter.name))
+                        } else {
+                            provided_file_path.clone()
+                        };
+
+                        let info = ChapterTemplateInfo {
+                            content: chapter.content.clone(),
+                            source: chapter_path.clone(),
+                            template_destination: destination.clone(),
+                        };
+                        (Some(info), destination)
+                    }
+                    None => (None, provided_file_path.clone()),
+                };
+
+                // Recursively call to get sub-chapter contents.
+                Some(
+                    get_subcontent_for_chapter(&chapter, new_path, provided_depth, depth + 1)
+                        .into_iter()
+                        .chain(chapter_info),
+                )
+            } else {
+                None
+            }
```

Comment:
Could this be turned into an early return? The [`let ... else` syntax](https://google.github.io/comprehensive-rust/control-flow/if-let-expressions.html) supports this in newer Rust versions (see the speaker notes there).